hathor/builder.py

```diff
@@ -36,11 +40,266 @@
+
+        manager = HathorManager(
+            reactor,
+            pubsub=pubsub,
+            consensus_algorithm=consensus_algorithm,
+            peer_id=peer_id,
+            tx_storage=tx_storage,
+            network=self._network,
+            wallet=wallet,
+            rng=self._rng,
+            checkpoints=self._checkpoints,
+            capabilities=self._capabilities,
+            environment_info=get_environment_info(self._cmdline, peer_id.id),
+            **kwargs
+        )
+
+        self.artifacts = BuildArtifacts(
+            settings=settings,
+            rng=self._rng,
+            reactor=reactor,
+            manager=manager,
+            pubsub=pubsub,
+            consensus=consensus_algorithm,
+            tx_storage=tx_storage,
+            indexes=indexes,
+            wallet=wallet,
+            rocksdb_storage=self._rocksdb_storage,
+        )
+
+        return self.artifacts
+
+    def set_event_manager(self, event_manager: EventManager) -> None:
+        self._event_manager = event_manager
+
+    def _get_peer_id(self) -> PeerId:
+        if self._peer_id is not None:
+            return self._peer_id
+        raise NotImplementedError
+
+    def set_rng(self, rng: Random) -> None:
+        self._rng = rng
+
+    def set_checkpoints(self, checkpoints: List[Checkpoint]) -> None:
+        self._checkpoints = checkpoints
+
+    def set_capabilities(self, capabilities: List[str]) -> None:
+        self._capabilities = capabilities
+
+    def set_peer_id(self, peer_id: PeerId) -> None:
+        self._peer_id = peer_id
+
+    def _get_settings(self) -> HathorSettingsType:
+        return HathorSettings()
+
+    def _create_pubsub(self, reactor: Reactor) -> PubSubManager:
+        if self._pubsub is not None:
+            return self._pubsub
+        return PubSubManager(reactor)
+
+    def _create_reactor(self) -> Reactor:
+        if self._reactor is not None:
+            return self._reactor
+        raise NotImplementedError
+
+    def _get_rocksdb_storage(self) -> RocksDBStorage:
+        assert self._rocksdb_path is not None
+        if self._rocksdb_storage is None:
+            self._rocksdb_storage = RocksDBStorage(path=self._rocksdb_path)
+        return self._rocksdb_storage
+
+    def _create_tx_storage(self) -> TransactionStorage:
+        if self._tx_storage is not None:
+            return self._tx_storage
+        if self._storage_type == StorageType.MEMORY:
+            return TransactionMemoryStorage()
+        elif self._storage_type == StorageType.ROCKSDB:
+            rocksdb_storage = self._get_rocksdb_storage()
+            use_memory_index = self._force_memory_index
+            return TransactionRocksDBStorage(rocksdb_storage, use_memory_indexes=use_memory_index)
+        raise NotImplementedError
+
+    def use_memory(self) -> None:
+        self._storage_type = StorageType.MEMORY
+
+    def use_rocksdb(self, path: str) -> None:
+        self._storage_type = StorageType.ROCKSDB
```

Comment:
Should it `assert self._storage_type is None` or if not an `assert` check and raise a `ValueError`?

Reply:
It might. But I don't see a problem with the following calls:

```python
builder = Builder()
builder.use_memory()
builder.use_rocksdb(path)
builder.use_memory()
artifacts = builder.build()
```

In this case, it would use memory objects.

Reply:
I think it makes sense, that is more consistent with how all the other methods work.